### PR TITLE
Convert `fetch-json` to Typescript

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

This converts `fetch-json` to TypeScript, as part of an effort to modernise the commercial code base. The benefit of `fetch-json` over the built-in `fetch` is that it throws an error if the response does not have an `ok` status code (200-299).

- Remove default export in favour of named export (ESLint complained)
- Handles malformed JSON, if `json()` is not a function on the response.

Follow-up from #23599, #23687 & #23644 .

## What is the value of this and can you measure success?

Typescript is how we type, nowadays.

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [X] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [X] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/lib/fetch-json.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import config from 'lib/config';
+
+const fetchJson = (
+	input: RequestInfo,
+	init: RequestInit = {},
+): Promise<unknown> => {
+	// #? we should use Object.assign for this assignment, but this currently breaks
+	// Karma tests that depend on fetch-json.js and have not been stubbed
```
Comment (mxdvl) [inline static/src/javascripts/lib/fetch-json.ts:8]: Are Karma tests precursors of Jest?

File: static/src/javascripts/lib/fetch-json.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	const options = init;
+	let path = typeof input === 'string' ? input : input.url;
+	if (!RegExp('^(https?:)?//').exec(path)) {
+		path = (config.get('page.ajaxUrl', '') as string) + path;
+		options.mode = 'cors';
+		options.credentials = 'include';
+	}
+
+	return fetch(path, options).then((resp) => {
```
Comment (mxdvl) [inline static/src/javascripts/lib/fetch-json.ts:17]: All calls to `fetch-json` have been checked for compliance to `(input: string, init?: object)`.

```suggestion
	let path = ''
	if (!RegExp('^(https?:)?//').exec(path)) {
		path = String(config.get('page.ajaxUrl', ''));
		init.mode = 'cors';
		init.credentials = 'include';
	}

	return fetch(path + input, init).then((resp) => {
```

**Review by mxdvl — COMMENTED** [review]
I think it’s worth using named export, so I’ll do just that.

File: static/src/javascripts/projects/common/modules/video/metadata.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import config from 'lib/config';
+import fetch from 'lib/fetch';
```
Comment (arelra) [inline static/src/javascripts/projects/common/modules/video/metadata.js:2]: Just out of interest, what was the issue here?

Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/video/metadata.js:2]: There needs to be a mock for node-fetch. It’ll come through the identity API conversion to Typescript.

**Review by arelra — APPROVED** [review]

**Review by shtukas — APPROVED** [review]

File: static/src/javascripts/lib/fetch-json.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	if (typeof resource !== 'string')
+		throw new Error('First argument should be of type `string`');
+
+	let path = '';
+	if (!RegExp('^(https?:)?//').exec(resource)) {
+		path = String(config.get('page.ajaxUrl', '')) + resource;
+		init.mode = 'cors';
+		init.credentials = 'include';
+	}
+	const resp = (await fetch(path, init)) as Response;
```
Comment (i-hardy) [inline static/src/javascripts/lib/fetch-json.ts:22]: If `resource` _does_ match the regex, `path` will never be assigned a new value, so the request will always be made with an empty string. This is breaking all uses of `fetchJson` that pass a full URL as `resource`.

Comment (mxdvl) [reply static/src/javascripts/lib/fetch-json.ts:22]: You’re right, I’ve raised a fix in #23693.

**Pull request merged**